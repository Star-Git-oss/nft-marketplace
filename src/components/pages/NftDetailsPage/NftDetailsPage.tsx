import React, { useEffect, useState } from 'react';
import axios from 'axios';
import { useTranslation } from 'react-i18next';
import { Helmet } from 'react-helmet';
import { EyeIcon, ShareIcon, HeartIcon } from '../../common/Icons/Icons';
import Col from '../../common/ui-library/Col/Col';
import Row from '../../common/ui-library/Row/Row';
import Image from '../../common/ui-library/Image/Image';
import Button from '../../common/ui-library/Button/Button';
import { H1, H4, SubTitle } from '../../common/Title/Title';
import { RoundedSpan, SquaredSpan } from '../../common/Shapes/Shapes';
import PurchaseModal from './PurchaseModal/PurchaseModal';
import NftImage from '../../common/NftImage/NftImage';
import { ApiProxyUri } from '../../../utils/utils';

type DetailsLabelProps = {
  label: string;
  value: string;
  image?: string;
};

const DetailsLabel: React.FC<DetailsLabelProps> = (props) => (
  
  <Row>
    <Col size="10">
      { props.image && 
      <Image rounded responsive src={props.image} />}
    </Col>
    <Col size="90">
      <H4 style={{ color: '#969393', margin: '0', fontWeight: 'bold' }}>
        {props.label}
      </H4>          
      <div style={{ fontWeight: 'bold', marginBottom: '20px' }}>
        {props.value}
      </div>
    </Col>
  </Row>

);

const NftDetailsPage: React.FC<LoadablePageType> = ( { setIsLoading }) => {

  // Modal for payment
  const [displayModal, setDisplayModal] = useState(false);

  // Nft information
  const [nft, setNft] = useState({} as NftObjectType);

  // Retrieve NFT info when component loaded
  useEffect(  () => {
    async function fetchData() {
      setIsLoading(true);
      const res = await axios.get(`${ApiProxyUri}/nft/1`);
      setNft(res.data.nft);
      setIsLoading(false);
    }
    fetchData();
  }, [setIsLoading]);

  const { t } = useTranslation();

  return (
    <>

      <Helmet>
        <title>{t('details.seo.title')}</title>
        <meta name="description" content={t('details.seo.description')} />
        <meta name="keywords" content={t('details.seo.keywords')} />
      </Helmet>

      {nft && (
        <>

          {displayModal && 
          <PurchaseModal 
            nft={nft} 
            closeModal={() => {setDisplayModal(false);}} 
          />}
          
          <Row>
            {/* NFT image  */}
            <Col small="100" medium="50" large="50">
              <div style={{ textAlign: 'center' }}>
                <NftImage 
                  full={true}
                  alt={nft.name} 
                  src={nft.image}
                />
              </div>
            </Col>

            {/* NFT details  */}
            <Col small="100" medium="50" large="50">
              <div style={{ margin: '0 auto' }}>
                <H1>{nft.name}</H1>
                <Button primary full onClick={() => {setDisplayModal(true);}}>
                  {t('details.buy')} - 
                  {nft.price}
                </Button>

                <SubTitle style={{ width: '100%', textAlign: 'center' }}>
                  Service fee 2.5%.
                </SubTitle>

                <div style={{ marginTop: '40px', height: '100px' }}>
                  <SquaredSpan><EyeIcon />
                    {nft.views}
                  </SquaredSpan>&nbsp;
                  <RoundedSpan><HeartIcon /></RoundedSpan>&nbsp;
                  <RoundedSpan><ShareIcon /></RoundedSpan>
                </div>

                <div>
                  <DetailsLabel 
                    label={t('details.owner')}
                    value={nft.owner}
                    image={nft.ownerPicture}
                  />
                  <DetailsLabel 
                    label={t('details.creator')}
                    value={nft.creator}
                    image={nft.creatorPicture}
                  />
                  <DetailsLabel 
                    label={t('details.type')}
                    value={nft.collectionName}
                  />
                </div>
              
              </div>
            </Col>
          </Row>
        </>
      )}
    </>
  );
};

export default NftDetailsPage;
