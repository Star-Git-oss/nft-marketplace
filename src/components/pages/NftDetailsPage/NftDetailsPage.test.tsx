/* eslint-disable no-console */
import React from 'react';
import axios from 'axios';
import { render, screen } from '@testing-library/react';
import renderer from 'react-test-renderer';
import NftDetailsPage from './NftDetailsPage';

// Known issue: error and warning does not work for console
// https://github.com/facebook/react/issues/7047
const original = console.error;

beforeEach(() => {
  console.error = jest.fn();
});

afterEach(() => {
  console.error = original;
});

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key) => key }),
}));

const nft = {
  nft: {
    'id':1,
    'labels': [1, 2, 3],
    'name': 'my NFT 1',
    'quantity': '10 of 10',
    'price': '0.2',
    'image': 'https://picsum.photos/200/300',
    'views': 5201,
    'owner': 'mickael canu',
    'creator': 'inri',
    'collectionName': 'Test name collection',
    'ownerPicture': 'owner.jpg',
    'creatorPicture': 'creator.jpg',
  }
};

// Axios mock
jest.mock('axios');
const resp = { data: nft };
axios.get.mockResolvedValue(resp);

describe('NftDetailsPage snapshots', () => {

  test('NftDetailsPage renders correctly', () => {
    const tree = renderer
      .create(<NftDetailsPage setIsLoading={false} />)
      .toJSON();
    expect(tree).toMatchSnapshot();
  });

});


it('renders component properly', () => {

  render(<NftDetailsPage setIsLoading={false} />);
  screen.findByText('details.buy');

  screen.findByText('0.2');

  // owner displayed
  screen.findByText('mickael canu');
  // creator displayed
  screen.findByText('inri');
  // owner picture displayed
  screen.findByText('owner.jpg');
  // creator picture displayed
  screen.findByText('creator.jpg');
  //  collection name displayed
  screen.findByText('Test name collection');
  // views displayed
  screen.findByText('5201');

    
});
