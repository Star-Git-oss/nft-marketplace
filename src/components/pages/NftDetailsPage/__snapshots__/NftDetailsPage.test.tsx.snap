// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NftDetailsPage snapshots NftDetailsPage renders correctly 1`] = `
<div
  className="sc-hKgILt fdNGpG"
>
  <div
    className="sc-dlfnbm dkRTsu"
  >
    <div
      style={
        Object {
          "textAlign": "center",
        }
      }
    >
      <img
        className="sc-hHftDr ciXPjr"
      />
    </div>
  </div>
  <div
    className="sc-dlfnbm dkRTsu"
  >
    <div
      style={
        Object {
          "margin": "0 auto",
        }
      }
    >
      <h1
        className="sc-gKsewC hqCRID"
      />
      <button
        className="sc-jSgupP cslRkz"
        onClick={[Function]}
      >
        details.buy
         -
      </button>
      <p
        className="sc-iqHYGH bzlcFs"
        style={
          Object {
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        Service fee 2.5%.
      </p>
      <div
        style={
          Object {
            "height": "100px",
            "marginTop": "40px",
          }
        }
      >
        <span
          className="sc-bqyKva eJtIgW"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 576 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M572.52 241.4C518.29 135.59 410.93 64 288 64S57.68 135.64 3.48 241.41a32.35 32.35 0 0 0 0 29.19C57.71 376.41 165.07 448 288 448s230.32-71.64 284.52-177.41a32.35 32.35 0 0 0 0-29.19zM288 400a144 144 0 1 1 144-144 143.93 143.93 0 0 1-144 144zm0-240a95.31 95.31 0 0 0-25.31 3.79 47.85 47.85 0 0 1-66.9 66.9A95.78 95.78 0 1 0 288 160z"
            />
          </svg>
        </span>
         
        <span
          className="sc-dQppl iCfrGj"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 512 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M458.4 64.3C400.6 15.7 311.3 23 256 79.3 200.7 23 111.4 15.6 53.6 64.3-21.6 127.6-10.6 230.8 43 285.5l175.4 178.7c10 10.2 23.4 15.9 37.6 15.9 14.3 0 27.6-5.6 37.6-15.8L469 285.6c53.5-54.7 64.7-157.9-10.6-221.3zm-23.6 187.5L259.4 430.5c-2.4 2.4-4.4 2.4-6.8 0L77.2 251.8c-36.5-37.2-43.9-107.6 7.3-150.7 38.9-32.7 98.9-27.8 136.5 10.5l35 35.7 35-35.7c37.8-38.5 97.8-43.2 136.5-10.6 51.1 43.1 43.5 113.9 7.3 150.8z"
            />
          </svg>
        </span>
         
        <span
          className="sc-dQppl iCfrGj"
        >
          <svg
            fill="currentColor"
            height="1em"
            stroke="currentColor"
            strokeWidth="0"
            style={
              Object {
                "color": undefined,
              }
            }
            viewBox="0 0 448 512"
            width="1em"
            xmlns="http://www.w3.org/2000/svg"
          >
            <path
              d="M352 320c-22.608 0-43.387 7.819-59.79 20.895l-102.486-64.054a96.551 96.551 0 0 0 0-41.683l102.486-64.054C308.613 184.181 329.392 192 352 192c53.019 0 96-42.981 96-96S405.019 0 352 0s-96 42.981-96 96c0 7.158.79 14.13 2.276 20.841L155.79 180.895C139.387 167.819 118.608 160 96 160c-53.019 0-96 42.981-96 96s42.981 96 96 96c22.608 0 43.387-7.819 59.79-20.895l102.486 64.054A96.301 96.301 0 0 0 256 416c0 53.019 42.981 96 96 96s96-42.981 96-96-42.981-96-96-96z"
            />
          </svg>
        </span>
      </div>
      <div>
        <div
          className="sc-hKgILt fdNGpG"
        >
          <div
            className="sc-dlfnbm lfGvUZ"
            size="10"
          />
          <div
            className="sc-dlfnbm hLdwEB"
            size="90"
          >
            <h4
              className="sc-pFZIQ bTNpbg"
              style={
                Object {
                  "color": "#969393",
                  "fontWeight": "bold",
                  "margin": "0",
                }
              }
            >
              details.owner
            </h4>
            <div
              style={
                Object {
                  "fontWeight": "bold",
                  "marginBottom": "20px",
                }
              }
            />
          </div>
        </div>
        <div
          className="sc-hKgILt fdNGpG"
        >
          <div
            className="sc-dlfnbm lfGvUZ"
            size="10"
          />
          <div
            className="sc-dlfnbm hLdwEB"
            size="90"
          >
            <h4
              className="sc-pFZIQ bTNpbg"
              style={
                Object {
                  "color": "#969393",
                  "fontWeight": "bold",
                  "margin": "0",
                }
              }
            >
              details.creator
            </h4>
            <div
              style={
                Object {
                  "fontWeight": "bold",
                  "marginBottom": "20px",
                }
              }
            />
          </div>
        </div>
        <div
          className="sc-hKgILt fdNGpG"
        >
          <div
            className="sc-dlfnbm lfGvUZ"
            size="10"
          />
          <div
            className="sc-dlfnbm hLdwEB"
            size="90"
          >
            <h4
              className="sc-pFZIQ bTNpbg"
              style={
                Object {
                  "color": "#969393",
                  "fontWeight": "bold",
                  "margin": "0",
                }
              }
            >
              details.type
            </h4>
            <div
              style={
                Object {
                  "fontWeight": "bold",
                  "marginBottom": "20px",
                }
              }
            />
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
